src/gurt/telemetry.c

```diff
@@ -2330,6 +2330,14 @@ d_tm_add_node(struct d_tm_node_t *src, struct d_tm_nodeList_t **nodelist)
 	}
 	return -DER_NOMEM;
 }
+
+/** create a unique key for this instance */
```

Comment:
I have a couple of thoughts about this.  One is that, to maintain unique keys that rely on getuid(), we should allocate some bits in the key to the uid itself, so that there is reduced chance of collision between the keys (perhaps 16 bits from the getuid and 16 bits from the base + instance?)  The second is that, the client consumer depends upon knowing the shared memory key for the engine instance that it is going to read telemetry from.  Currently, that key is deterministic, as it is the sum of D_TM_SHARED_MEMORY_KEY + srv_idx.  If the key now also depends on the uid of the calling process that created it, then it looks like only a consumer with the same uid as the producer would be able to access the shared memory segment.  That behavior seems contrary to the title of the PR.  Is the intent to make access to the shared memory segment more restrictive than the 0660 permissions make it today?

Reply:
Ok, I see what you mean, thanks.

For "multiple users", the commit msg did not say it clear enough. It was to fix the problem when multiple users start daos_servers on same set of nodes, that usage was supported before (for example two users shard the same set of testing nodes). In that case, if we generate the key only by "D_TM_SHARED_MEMORY_KEY + srv_idx" that is not enough. So this patch just add getuid() to the key.
On the other hand, I am not very sure but it seems reasonable to me that require the user to run "daos_metrics" should be same user who started the servers? This align with the current "0660" permissions. But if we need to support the case for example root user starts servers, and allow other users to run daos_metrics, that may need to add another option to daos_metrics binary to make it aware that case, and change "0660" to "0666". What do you think?

Agree that "to maintain unique keys that rely on getuid(), we should allocate some bits in the key to the uid itself". That probably can be refined by other patch later.

Reply:
I wasn't aware of the use model of running multiple instances of the control plane on a single machine.  That does require a different scheme to generate the shared memory key and make that deterministic so that the consumer can operate.  One option would be to add a unique shared memory base key to the daos_server.yml, which would be passed to the daos_engine so that the engines started by that control plane instance would have unique shared memory segments from the engines started by another control plane instance, provided the daos_server.yml is actually updated to some unique value for each control plane instance.  The trouble is, that solution requires the developer who launched an instance to tweak their configuration file, rather that just use a default value.  That creates a new opportunity to get it wrong and cause startup to fail due to the collision of default keys.  On the plus side, the daos_server reads the yml config and knows the key, so it will know what it is when it executes the consumer built into the control plane.  On the negative side, daos_metrics which is a stand-alone program, will need to be told the shared memory key that matches the daos_server instance that was intended.  That's not a big deal though.

There was a desire to enforce that daos_metrics be in the same group as the daos_engine, so that daos_metrics would have access to the shared memory, and exclude the rest.  This implies that if you start the daos_engine as yourself, your user should be part of the same group that was setup for the daos_engine / daos_metrics.  

@mjmac what are your thoughts on this?

Reply:
First of all, I think we should keep the coverity defects separate from changes to the telemetry library.

With regard to the permissions, the reason we cannot make the shared memory segment world-writable is that it opens a security hole and could result in engine crashes if the telemetry segment gets corrupted.

Running multiple control plane servers concurrently on the same storage host is not a use case that we support for production, and I am reluctant to make changes to production code in order to support developer use cases, because then we have to maintain and test that code.

Are you actually trying to run > 1 daos_server process at the same time, or is this just a problem where one developer's environment needs to be cleaned up before another one can use the nodes? If the latter case, then I suggest just writing a script to clean things up.

Reply:
> One option would be to add a unique shared memory base key to the daos_server.yml, which would be passed to the daos_engine so that the engines started by that control plane instance would have unique shared memory segments
Or may reuse existing unique config like "fabric_iface_port".

> Are you actually trying to run > 1 daos_server process at the same time
When 2 users share the same node and run test concurrently, it is this case. I am not quite sure if it is only a "developer use case" although, it can work before (in history it broken occasionally but a few time later can work again).

Since the use case is not officially supported now, I can workaround it by other method. will take out other code and close this PR.

Reply:
the added dbg log looks useful, so reopen and resubmit a simpler small patch.